### ISSUE TITLE
Implementation of Data Value Embedding (DVEmb)

### ISSUE THREAD

**Issue by BilyHurington** [TRAIS-Lab/dattri#191]
Implement DVEmb in https://arxiv.org/pdf/2412.09538.

1. what methods (external methods, i.e., the ones that will be directly called by users) will I add/modify. Use docstring directly to answer the question.

I plan to implement a new attributor class DVEmbAttributor, including the following methods.

```py
def __init__(
        self,
        model: nn.Module,
        loss_func: Callable,
        device: str = "cpu",
        proj_dim: Optional[int] = None,
        factorization_type: str = "none",
    ) -> None:
        """Initializes the DVEmb attributor.

        Args:
            model: The PyTorch model to be attributed.
            loss_func: A per-sample loss function. It should take model parameters
                       and a single data sample (unbatched) as input.
            device: The device to run computations on (e.g., "cpu", "cuda").
            proj_dim: The dimension for projection (if used).
            factorization_type: Type of gradient factorization to use. Options are
                                "none"(default),
                                "kronecker"(same with paper),
                                or "elementwise"(more memory-efficient).
        """
```

```py
def cache_gradients(
        self,
        epoch: int,
        batch_data: tuple[Tensor, ...],
        indices: Tensor,
        learning_rate: float,
    ) -> None:
        """Cache per-sample gradients for a specific epoch and training step.

        Args:
            epoch: The current epoch number.
            batch_data: A tuple containing the batch of inputs and targets,
                        e.g., (inputs, labels).
            indices: A tensor containing the original indices for the samples
                     in batch_data.
            learning_rate: The learning rate for this step.
        """
```

```py
def compute_embeddings( 
        self,
        gradients: Optional[Dict[int, List[Tensor]]] = None,
        learning_rates: Optional[Dict[int, List[float]]] = None,
    ) -> None:
        """Computes data value embeddings for each epoch separately.

        Args:
            gradients: Optional external gradients instead of cached ones
                (e.g., (epoch -> list of per-sample gradients)).
            learning_rates: Optional external learning rates instead of cached ones
                (e.g., (epoch -> list of learning rates)).
        """
```

```py
def attribute(
        self,
        test_dataloader: DataLoader,
        epoch: Optional[int] = None,
        traindata_indices: Optional[List[int]] = None,
    ) -> torch.Tensor:
        """Calculates influence scores for a test set for one or all epochs.

        Args:
            test_dataloader: A dataloader for the test set.
            epoch: Optional. If specified, returns scores using embeddings from that
                   epoch. If None, returns scores based on the sum of embeddings
                   across all epochs.
            traindata_indices: Optional. A list of training sample indices for which
                               to compute influence. If None, computes for all.
        """
```

2. How will users use this function? Please use a very simple (but comprehensive) example call to answer the question.

```py
task = AttributionTask(model=model, loss_fn=loss_fn)

dvemb_attributor = DVEmbAttributor(
    model=model,
    loss_func=criterion,
    device="cuda",
    proj_dim=4096,
    factorization_type="kronecker",
)

# Inside a training loop
for epoch in range(num_epochs):
    for i, batch in enumerate(train_loader):
        # ... standard training forward pass and optimizer step ...
        
        # Store gradient info for the current batch
        dvemb_attributor.cache_gradients(
            epoch=epoch,
            iteration=i,
            batch_data=batch,
            learning_rate=optimizer.param_groups[0]['lr']
        )

dvemb_attributor.compute_embeddings()

influence_scores = dvemb_attributor.attribute(test_loader)
```

3. How will I implement it?

- 1. Implement `__init__` to configure the attributor and `cache_gradients` to efficiently cache decomposed and optionally projected gradient components during each training iteration.
- 2. Implement `compute_embeddings` to execute a backpropagation algorithm that processes the cached gradients in reverse, computing the final Data Value Embedding for each training point on a per-layer basis.
- 3. Implement `attribute` to calculate influence scores by loading the pre-computed embeddings and taking their dot product with the gradients of the test data.

**Comment by TheaperDeng** [comment]
Yes you are right, if we have already got the training gradient as input, we don't need the training loader again for the `compute_embeddings`. As for the `traindata_indices` in the `attribute` I think we can follow your design here right now. Though it seems to be a little bit different from others.

The design LGTM right now.
